chasm/lib/scheduler/invoker_tasks.go

```diff
@@ -0,0 +1,603 @@
+func (e *InvokerExecuteTaskExecutor) terminateWorkflows(
+	ctx invokerTaskExecutorContext,
+	logger log.Logger,
+	scheduler *Scheduler,
+	targets []*commonpb.WorkflowExecution,
+) (result executeResult) {
+	for _, wf := range targets {
+		if !ctx.takeNextAction() {
+			break
+		}
+
+		err := e.terminateWorkflow(ctx, scheduler, wf)
+		if err != nil {
+			logger.Error("Failed to terminate workflow", tag.Error(err), tag.WorkflowID(wf.WorkflowId))
+			e.MetricsHandler.Counter(metrics.ScheduleTerminateWorkflowErrors.Name()).Record(1)
+		}
+
+		// Terminates are only attempted once.
+		result.CompletedTerminates = append(result.CompletedTerminates, wf)
+	}
+	return
+}
+
+// startWorkflows executes the provided list of starts, returning a result with their outcomes.
+func (e *InvokerExecuteTaskExecutor) startWorkflows(
+	ctx invokerTaskExecutorContext,
+	logger log.Logger,
+	scheduler *Scheduler,
+	starts []*legacyschedulespb.BufferedStart,
+) (result executeResult, startResults []*schedulepb.ScheduleActionResult) {
+	metricsWithTag := e.MetricsHandler.WithTags(
+		metrics.StringTag(metrics.ScheduleActionTypeTag, metrics.ScheduleActionStartWorkflow))
+
+	for _, start := range starts {
+		// Starts that haven't been executed yet will remain in `BufferedStarts`,
+		// without change, so another ExecuteTask will be immediately created to continue
+		// processing in a new task.
+		if !ctx.takeNextAction() {
+			break
+		}
+
+		startResult, err := e.startWorkflow(ctx, scheduler, start)
+		if err != nil {
+			logger.Error("Failed to start workflow", tag.Error(err))
+
+			// Don't count "already started" for the error metric or retry, as it is most likely
+			// due to misconfiguration.
+			if !isAlreadyStartedError(err) {
+				metricsWithTag.Counter(metrics.ScheduleActionErrors.Name()).Record(1)
+			}
+
+			if isRetryableError(err) {
+				// Apply backoff to start and retry.
+				e.applyBackoff(start, err)
+				result.RetryableStarts = append(result.RetryableStarts, start)
+			} else {
+				// Drop the start from the buffer.
+				result.FailedStarts = append(result.FailedStarts, start)
+			}
+
+			continue
+		}
+
+		metricsWithTag.Counter(metrics.ScheduleActionSuccess.Name()).Record(1)
+		result.CompletedStarts = append(result.CompletedStarts, start)
+		startResults = append(startResults, startResult)
+	}
+	return
+}
+
+func (e *InvokerProcessBufferTaskExecutor) Validate(
+	ctx chasm.Context,
+	invoker *Invoker,
+	attrs chasm.TaskAttributes,
+	_ *schedulespb.InvokerProcessBufferTask,
+) (bool, error) {
+	return validateTaskHighWaterMark(invoker.GetLastProcessedTime(), attrs.ScheduledTime)
+}
+
+func (e *InvokerProcessBufferTaskExecutor) Execute(
+	ctx chasm.MutableContext,
+	invoker *Invoker,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerProcessBufferTask,
+) error {
+	scheduler, err := invoker.Scheduler.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %w",
+			serviceerror.NewInternal("Failed to read component"),
```

Comment:
Inside a transaction it's probably fine for now to treat this as a internal error since all component are loaded. 
If we do partial load in the future, ^ assumption may not hold.